lib/cdo/analytics/milestone_parser.rb

```diff
@@ -0,0 +1,118 @@
+    cache_file = MILESTONE_CACHE_V2
+    File.copy(MILESTONE_CACHE, cache_file) unless File.file?(cache_file)
+    cache = File.file?(cache_file) ? JSON.parse(IO.read(cache_file)) : {}
+    parser = self.new(cache, AWS::S3::connect_v2!)
+    parser.count.tap{|_|IO.write MILESTONE_CACHE_V2, JSON.pretty_generate(parser.cache)}
+  end
+
+  def initialize(cache, s3_client)
+    @cache = cache
+    @s3_client = s3_client
+    @s3_resource = Aws::S3::Resource.new(client: s3_client)
+  end
+
+  # Parses all milestone logs in s3://cdo-logs/hosts/**/dashboard/milestone.log*
+  # Return total lines of code count
+  def count
+    start_time = Time.now
+    debug 'Fetching milestone logs..'
+
+    hosts = s3_client.list_objects(bucket: 'cdo-logs', prefix: 'hosts/', delimiter: '/').data.common_prefixes.map(&:prefix)
+    hosts.select! do |prefix|
+      match = prefix.match /^hosts\/(?<host>[^\/]+)\//
+      match && !(IGNORE_HOSTS.include? match[:host])
+    end
+    logs = Parallel.map(hosts, in_threads: 16) do |host|
+      s3_resource.bucket('cdo-logs').objects(prefix: "#{host}dashboard/milestone.log").to_a
+    end.flatten
+    debug "Found #{logs.length} logs.."
+    counts = logs.map do |log|
+      (cache[log.key] = count_lines_of_code(log))['count']
+    end
+    total = counts.reduce(:+)
+    debug "Finished processing (#{(Time.now - start_time).round(2)}s), total count: #{total}"
+    total
+  end
+
+  def count_lines_of_code(log)
+    length = fetch_bytes = log.size
+    return {'count' => 0} if length == 0
+    key = log.key
+    filename = File.basename(key)
+    ext = File.extname(filename)
+    raise ArgumentError, "Don't know how to process #{ext} files." unless %w(.log .gz).include? ext
+    path = pegasus_dir 'cache', filename
+
+    # Remove surrounding quotation marks from etag string
+    etag = log.etag.gsub(/\A"|"\Z/, '')
+    partial_count = 0
+    md5 = nil
+    fetch_params = {response_target: path}
+
+    if (cached = cache[key]).is_a?(Hash)
+      return cached if etag == cached['etag']
+      if cached['md5']
+        compare_bytes = [COMPARE_BYTE_LENGTH, length, cached['length']].min
+        debug "#{key} modified, comparing starting #{compare_bytes} bytes.."
+        check_value = log.get(range: "bytes=0-#{compare_bytes - 1}").body.read
+        md5 = Digest::MD5.hexdigest(check_value)
+        if cached['md5'] == md5
+          partial_count = cached['count']
+          cached_length = cached['length']
+          fetch_bytes = (length - cached_length)
+          fetch_params.merge!(range: "bytes=#{cached_length}-")
+          debug "Starting content match, downloading remaining #{fetch_bytes} bytes.."
+        end
+      end
+    else
+      # Assume cached .gz files without etag haven't been modified
+      return {'count' => cached, 'etag' => etag} if ext == '.gz' && cached.is_a?(Numeric)
+    end
+
+    debug "Downloading #{key}, #{fetch_bytes} bytes.."
+    log.get fetch_params
+    stub_fetch(key, path, fetch_bytes)
```

Comment:
yeah, just a testing hook, not ideal- probably cleaner to call a method that actually does something useful, and then override that in the test class